stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+   * times, { t1, t2, t3, ... } using the stiff backward differentiation formula
+   * (BDF) solver in CVODES.
+   *
+   * @return std::vector of Eigen::Matrix of the states of the ODE, one for each
+   *   solution time (excluding the initial state)
+   */
+  std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> operator()() {
+    const double t0_dbl = value_of(t0_);
+    const auto ts_dbl = value_of(ts_);
+    std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> y_return(
+        ts_.size());
+
+    double t_init = t0_dbl;
+    for (size_t n = 0; n < ts_dbl.size(); ++n) {
+      double t_final = ts_dbl[n];
+      if (t_final != t_init) {
+        if (is_var_return) {
+          int ncheck;
+
+          int error_code
+              = CVodeF(solver_->cvodes_mem_, t_final,
+                       solver_->nv_state_forward_, &t_init, CV_NORMAL, &ncheck);
+
+          if (error_code == CV_TOO_MUCH_WORK) {
+            throw_domain_error(solver_->function_name_, "", t_final,
+                               "Failed to integrate to next output time (",
+                               ") in less than max_num_steps steps");
+          } else {
+            check_flag_sundials(error_code, "CVodeF");
+          }
+
+        } else {
+          int error_code
+              = CVode(solver_->cvodes_mem_, t_final, solver_->nv_state_forward_,
+                      &t_init, CV_NORMAL);
+
+          if (error_code == CV_TOO_MUCH_WORK) {
+            throw_domain_error(solver_->function_name_, "", t_final,
+                               "Failed to integrate to next output time (",
+                               ") in less than max_num_steps steps");
+          } else {
+            check_flag_sundials(error_code, "CVode");
+          }
+        }
+      }
+      store_state(n, state_forward_, y_return[n]);
+
+      t_init = t_final;
+    }
+
+    forward_returned_ = true;
+    return y_return;
+  }
+
+  virtual void chain() {
+    if (solver_->cvodes_mem_ == nullptr) {
+      return;
+    }
+
+    if (!forward_returned_) {
+      return;
+    }
+
+    if (!is_var_return) {
+      return;
+    }
+
+    // for sensitivities wrt to ts we do not need to run the backward
+    // integration
+    if (is_var_ts) {
+      for (int i = 0; i < ts_.size(); ++i) {
+        Eigen::VectorXd step_sens = Eigen::VectorXd::Zero(N_);
+        for (int j = 0; j < N_; ++j) {
+          step_sens.coeffRef(j) += non_chaining_varis_[i * N_ + j]->adj_;
+        }
+
+        ts_varis_[i]->adj_ += step_sens.dot(
+            rhs(value_of(ts_[i]), solver_->y_[i], value_of_args_tuple_));
+      }
+
+      if (is_var_only_ts) {
+        return;
+      }
+    }
+
+    state_backward_.setZero();
+    quad_.setZero();
+
+    // At every time step, collect the adjoints from the output
+    // variables and re-initialize the solver
+    double t_init = value_of(ts_.back());
+    for (int i = ts_.size() - 1; i >= 0; --i) {
```

Comment:
[optional] We know the actual type here is an `Eigen::Index` 

```suggestion
    for (Eigen::Index i = ts_.size() - 1; i >= 0; --i) {
```

Reply:
but ts_ is  std::vector, so it's indexed by std::size_t I suppose, no?

Reply:
Oh good point! Thought it was an eigen vector